Batteries/Data/List/Lemmas.lean

```diff
@@ -482,6 +491,83 @@ theorem Sublist.erase_diff_erase_sublist {a : α} :
 
 end Diff
 
+/-! ### prefix, suffix, infix -/
+
+theorem singleton_prefix_cons (a) : [a] <+: a :: l :=
+  (prefix_cons_inj a).mpr nil_prefix
+
+theorem ne_nil_of_not_prefix (h : ¬l₁ <+: l₂) : l₁ ≠ [] := by
+  intro heq
+  simp [heq, nil_prefix] at h
+
+section
+
+variable [DecidableEq α]
+
+theorem commonPrefix_prefix_left (l₁ l₂ : List α) : commonPrefix l₁ l₂ <+: l₁ := by
+  match l₁, l₂ with
+  | [],   _  => simp [commonPrefix]
+  | _::_, [] => simp [commonPrefix]
+  | a₁::l₁, a₂::l₂ =>
+    simp only [commonPrefix]
+    cases Decidable.em (a₁ = a₂)
+    · next h =>
+      simp only [h, ↓reduceIte, cons_prefix_cons, true_and]
+      apply commonPrefix_prefix_left
+    · next h =>
+      simp [h]
```

Comment:
I don't usually see `cases Decidable.em` in code. Instead I'd see `by_cases`:
```suggestion
    by_cases h : a₁ = a₂
    · simp only [h, ↓reduceIte, cons_prefix_cons, commonPrefix_prefix_left, and_self]
    · simp [h]
```
or `split` since there's an `if`:
```suggestion
    split
    next h =>
      simp only [h, cons_prefix_cons, true_and, commonPrefix_prefix_left]
    next h =>
      simp [h]
```

Reply:
In mathlib we'd often use `obtain rfl | hxy := Decidable.eq_or_ne x y`; perhaps that should be upstreamed to batteries?

Reply:
> I don't usually see `cases Decidable.em` in code. Instead I'd see `by_cases`:

It opens the `Classical` namespace, doesn't it? The `simp` tactic have already used classical theorems, anyway.